src/main/scala/edu/alexy/lectures/concurrency/RoboAdviser.scala

```diff
@@ -4,48 +4,113 @@ import edu.alexy.lectures.concurrency.model.Company
+  def getAllTickersRetryable(retries: Int = 10): Future[Seq[String]] =
+    futureRetry(Db.getAllTickers, retries).transform {
+      case Failure(exception) => Failure(DbError)
+      case other              => other
+    }
 
   // Task 3.
   // Implement a fallback strategy for 'Db.getCompanyLastFinancials'.
   // 'Db.getCompanyLastFinancials' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getCompanyRetryable(ticker: String, retries: Int = 10): Future[Option[Company]] = ???
+  def getCompanyRetryable(ticker: String,
+                          retries: Int = 10): Future[Option[Company]] =
+    futureRetry(Db.getCompanyLastFinancials(ticker), retries).transform {
+      case Failure(exception) => Failure(DbError)
+      case other              => other
+    }
 
   // Task 4.
   // Implement a fallback strategy 'WebApi.getPrice'.
   // 'WebApi.getPrice' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] = ???
+  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] =
+    futureRetry(WebApi.getPrice(ticker), retries).transform {
+      case Failure(exception) => Failure(WebApiError)
+      case other              => other
+    }
 
   // Task 5.
   // Using retryable functions return all tickers with their real time prices.
-  def getAllTickersPrices: Future[Seq[(String, Double)]] = ???
+  def getAllTickersPrices: Future[Seq[(String, Double)]] =
+    for (tickers <- getAllTickersRetryable();
+         tikerWithPrice <- Future.traverse(tickers) { ticker =>
+           Future.successful(ticker).zip(getPriceRetryable(ticker))
+         }) yield tikerWithPrice
 
   // Task 6.
   // Using `getCompanyRetryable` and `getPriceRetryable` functions return a company with its real time stock price.
-  def getCompanyFinancialsWithPrice(ticker: String): Future[(Company, Double)] = ???
+  def getCompanyFinancialsWithPrice(ticker: String): Future[(Company, Double)] = {
+    getCompanyRetryable(ticker).collect {
+      case Some(company) =>
+        Future.successful(company).zip(getPriceRetryable(company.ticker))
+    }
+  }.flatten
 
   // Task 7.
   // Implement a function that returns a list of chip ('Company.isCheap') companies
   // with their real time stock prices using 'getAllTickersRetryable' and
   // 'getCompanyFinancialsWithPrice' functions.
-  def buyList: Future[Seq[(Company, Double)]] = ???
+  def buyList: Future[Seq[(Company, Double)]] = {
+    for (tickers <- getAllTickersRetryable();
+         companyWithPriceSeq <- Future.traverse(tickers) { ticker =>
+           getCompanyFinancialsWithPrice(ticker)
+             .collect {
+               case companyWithPrice @ (company, price)
+                   if company.isCheap(price) =>
+                 Success(companyWithPrice)
```

Comment:
You don't have to use `Try` projection here. Just a value.
Also, filtering of cheap companies can be done on a new line inside the for-comp:
`result = companyWithPriceSeq.filter { case (c, p) => c.isCheap(p) }`